peer/abstractlist/list.go

```diff
@@ -0,0 +1,592 @@
+	default:
+	}
+}
+
+// waitForPeerAddedEvent waits until a peer is added to the peer list or the
+// given context finishes.
+// Must NOT be run in a mutex.Lock()
+func (pl *List) waitForPeerAddedEvent(ctx context.Context) error {
+	if _, ok := ctx.Deadline(); !ok {
+		return pl.newNoContextDeadlineError()
+	}
+
+	select {
+	case <-pl.peerAvailableEvent:
+		return nil
+	case <-ctx.Done():
+		return pl.newUnavailableError(ctx.Err())
+	}
+}
+
+func (pl *List) newNoContextDeadlineError() error {
+	return yarpcerrors.Newf(yarpcerrors.CodeInvalidArgument, _noContextDeadlineError, pl.name)
+}
+
+func (pl *List) newUnavailableError(err error) error {
+	return yarpcerrors.Newf(yarpcerrors.CodeUnavailable, _unavailableError, pl.name, err.Error())
+}
+
+// NumAvailable returns how many peers are available.
+func (pl *List) NumAvailable() int {
+	pl.lock.RLock()
+	defer pl.lock.RUnlock()
+
+	num := 0
+	for _, p := range pl.peers {
+		if p.status.ConnectionStatus == peer.Available {
+			num++
+		}
+	}
+	return num
+}
+
+// NumUnavailable returns how many peers are unavailable while the list is
+// running.
+func (pl *List) NumUnavailable() int {
+	pl.lock.RLock()
+	defer pl.lock.RUnlock()
+
+	num := 0
+	for _, p := range pl.peers {
+		if p.status.ConnectionStatus == peer.Unavailable {
+			num++
+		}
+	}
+	return num
+}
+
+// NumUninitialized returns how many peers are unavailable because the peer
+// list was stopped or has not yet started.
+func (pl *List) NumUninitialized() int {
+	pl.lock.RLock()
+	defer pl.lock.RUnlock()
+
+	return len(pl.offlinePeers)
+}
+
+// Available returns whether the identifier peer is available for traffic.
+func (pl *List) Available(pid peer.Identifier) bool {
+	pl.lock.RLock()
+	defer pl.lock.RUnlock()
+
+	if p, ok := pl.peers[pid.Identifier()]; ok {
+		return p.status.ConnectionStatus == peer.Available
+	}
+	return false
+}
+
+// Uninitialized returns whether the identifier peer is present but uninitialized.
+func (pl *List) Uninitialized(pid peer.Identifier) bool {
+	pl.lock.RLock()
+	defer pl.lock.RUnlock()
+
+	_, exists := pl.offlinePeers[pid.Identifier()]
+	return exists
+}
+
+// Peers returns a snapshot of all retained (available and
+// unavailable) peers.
+func (pl *List) Peers() []peer.StatusPeer {
+	pl.lock.RLock()
+	defer pl.lock.RUnlock()
+
+	peers := make([]peer.StatusPeer, 0)
+	for _, p := range pl.peers {
+		peers = append(peers, p.peer)
+	}
+	return peers
+}
+
+func (pl *List) onlinePeerIdentifiers() []peer.Identifier {
+	addrs := make([]string, 0, len(pl.peers))
+	for addr := range pl.peers {
+		addrs = append(addrs, addr)
+	}
+	sort.Strings(addrs)
+
+	ids := make([]peer.Identifier, len(addrs))
+	for i, addr := range addrs {
+		ids[i] = pl.peers[addr].peer
+	}
+	return ids
+}
+
+func (pl *List) offlinePeerIdentifiers() []peer.Identifier {
+	addrs := make([]string, 0, len(pl.offlinePeers))
+	for addr := range pl.offlinePeers {
+		addrs = append(addrs, addr)
+	}
+	sort.Strings(addrs)
```

Comment:
It could, but it would be unique. I’ve added notes to clarify that these two functions are not copypasta, just very superficially similar.

Reply:
sorry I should have used the multi-select, I just meant the code from `sort.Strings(addrs)` to the end of the function is the same

Reply:
The code from sort.Strings(addrs) down is also subtly different. There’s also not a possibility of generalizing it because the map types are different. Hash tag generics.